File: bem.py
#print('hello chaman ') FOSSEE 2025 @NIT_kurukshetra


class bm_Data:
    def __init__(self): 
        self.L=float(input("Enter length of the beam in metre: "))
        self.w1=float(input("Enter load w1 in KN: "))
        self.w2=float(input("Enter load w2 in KN: "))
        self.x=float(input("Enter distance between w1 and w2: "))


beamdata=bm_Data()

def Reacn(beamdata,a=0):
    ord_1=(beamdata.L-a)/beamdata.L
    ord_2=(beamdata.L-(a+beamdata.x))/beamdata.L
    Rx_A = beamdata.w1*ord_1+beamdata.w2*ord_2
    Rx_B = beamdata.w1+ beamdata.w2-Rx_A #1 more way
    return Rx_A,Rx_B


def max_Rxn(beamdata):
    max_rxa=0.0
    max_rxb=0.0
   
    for a in range (0,int(beamdata.L+1)):
       
        Rx_A,Rx_B=Reacn(beamdata,a)
        if Rx_A>max_rxa and Rx_A<=(beamdata.w1+beamdata.w2):
            max_rxa=Rx_A
            pos_a=a
        
        if Rx_B>max_rxb and Rx_B<=(beamdata.w1+beamdata.w2): 
            max_rxb=Rx_B
            pos_b=a
                
    print(f"\nMaximum Rx_A: {max_rxa:.2f} KN @ w1 at {pos_a} m ")
    print(f"Maximum Rx_B: {max_rxb:.2f} KN @ w1 at {pos_b} m ")

max_Rxn(beamdata)


def BM_01(beamdata,s):
    if s<beamdata.x:
        bm_ord2=(beamdata.L-beamdata.x)*(s*(beamdata.L-s))/(beamdata.L*beamdata.L)
    else:   
     bm_ord2=(beamdata.L-s)*beamdata.x/beamdata.L
    bm = beamdata.w1*0+beamdata.w2*bm_ord2
    return round(bm,2)

lis_BM_01=[]
for  s in range(0,int(beamdata.L+1)):
    lis_BM_01.append(BM_01(beamdata,s))
    
print(f"BM_01 at w1=0 m is: {lis_BM_01}")


def SF_01(beamdata,t):  

    ord_1= t/beamdata.L
    ord_2= (t+beamdata.x)/beamdata.L

    if (t)<0.5*beamdata.L:
         Sf=  -(beamdata.w1*ord_1 + beamdata.w2*ord_2)
    elif (t>0.5*beamdata.L and t+beamdata.x<=beamdata.L): 
         Sf= (beamdata.w2*(1-ord_1) + beamdata.w1*(1-ord_2))
    else:
        Sf=((-beamdata.w1*(ord_2) + beamdata.w2*(1-ord_1)))

    return round(Sf,2)


lis_SF_01=[]
for t in range(0,int(beamdata.L+1-beamdata.x)):
    lis_SF_01.append((SF_01(beamdata,t)))

print("SF_01 at mid-span is:",lis_SF_01)   



def max_BM(beamdata):
    max_val=0.00
    for y in range(0,int(beamdata.L+1)): 
        ord_1= y*(beamdata.L-y)/beamdata.L
        ord_2= (y+beamdata.x)*(beamdata.L-(y+beamdata.x))/beamdata.L
        Bm_y= (beamdata.w1*ord_1 + beamdata.w2*ord_2)
        

        if Bm_y>max_val:
            max_val=round(Bm_y,2)
            pos_y=y
        
            
    print(f"max BM is {max_val:.2f}  kN-m and position y from left support is {pos_y} m\n")


 

def max_SF(beamdata):
    max_val=0.00
    for z in range(0,int(beamdata.L+1)): 
        ord_1= z/beamdata.L
        ord_2= (z+beamdata.x)/beamdata.L
        if z>0.5*beamdata.L:
         Sf_z= (beamdata.w1*ord_1 + beamdata.w2*ord_2)
        else:
         Sf_z= (beamdata.w2*ord_1 + beamdata.w1*ord_2)
       

        if Sf_z>max_val:
            max_val=Sf_z
            pos_z=z
    print(f"\nmax SF is {max_val:.2f} kN and position z from left support is {pos_z:} m")

max_SF(beamdata)
max_BM(beamdata)
    

File: bem.js
class BeamData {
  constructor(L, w1, w2, x) {
    this.L = parseFloat(L);
    this.w1 = parseFloat(w1);
    this.w2 = parseFloat(w2);
    this.x = parseFloat(x);
  }
}

function calculateReactions(beamData, a = 0) {
  const ord1 = (beamData.L - a) / beamData.L;
  const ord2 = (beamData.L - (a + beamData.x)) / beamData.L;
  const RxA = beamData.w1 * ord1 + beamData.w2 * ord2;
  const RxB = beamData.w1 + beamData.w2 - RxA;
  return { RxA, RxB };
}

function findMaxReactions(beamData) {
  let maxRxA = 0.0,
    maxRxB = 0.0,
    posA = 0,
    posB = 0;

  for (let a = 0; a <= beamData.L; a++) {
    const { RxA, RxB } = calculateReactions(beamData, a);
    if (RxA > maxRxA && RxA <= beamData.w1 + beamData.w2) {
      maxRxA = RxA;
      posA = a;
    }
    if (RxB > maxRxB && RxB <= beamData.w1 + beamData.w2) {
      maxRxB = RxB;
      posB = a;
    }
  }

  return {
    maxRxA: maxRxA.toFixed(2),
    posA,
    maxRxB: maxRxB.toFixed(2),
    posB,
  };
}

function calculateBM(beamData, s) {
  let bmOrd2;
  if (s < beamData.x) {
    bmOrd2 =
      ((beamData.L - beamData.x) * (s * (beamData.L - s))) /
      (beamData.L * beamData.L);
  } else {
    bmOrd2 = ((beamData.L - s) * beamData.x) / beamData.L;
  }
  const bm = beamData.w1 * 0 + beamData.w2 * bmOrd2;
  return parseFloat(bm.toFixed(2));
}

function calculateSF(beamData, t) {
  const ord1 = t / beamData.L;
  const ord2 = (t + beamData.x) / beamData.L;

  let Sf;
  if (t < 0.5 * beamData.L) {
    Sf = -(beamData.w1 * ord1 + beamData.w2 * ord2);
  } else if (t > 0.5 * beamData.L && t + beamData.x <= beamData.L) {
    Sf = beamData.w2 * (1 - ord1) + beamData.w1 * (1 - ord2);
  } else {
    Sf = -beamData.w1 * ord2 + beamData.w2 * (1 - ord1);
  }

  return parseFloat(Sf.toFixed(2));
}

function findMaxBM(beamData) {
  let maxVal = 0.0,
    posY = 0;

  for (let y = 0; y <= beamData.L; y++) {
    const ord1 = (y * (beamData.L - y)) / beamData.L;
    const ord2 =
      ((y + beamData.x) * (beamData.L - (y + beamData.x))) / beamData.L;
    const BmY = beamData.w1 * ord1 + beamData.w2 * ord2;

    if (BmY > maxVal) {
      maxVal = parseFloat(BmY.toFixed(2));
      posY = y;
    }
  }

  return { maxBM: maxVal.toFixed(2), posY };
}

function findMaxSF(beamData) {
  let maxVal = 0.0,
    posZ = 0;

  for (let z = 0; z <= beamData.L; z++) {
    const ord1 = z / beamData.L;
    const ord2 = (z + beamData.x) / beamData.L;
    let SfZ;

    if (z > 0.5 * beamData.L) {
      SfZ = beamData.w1 * ord1 + beamData.w2 * ord2;
    } else {
      SfZ = beamData.w2 * ord1 + beamData.w1 * ord2;
    }

    if (SfZ > maxVal) {
      maxVal = SfZ;
      posZ = z;
    }
  }

  return { maxSF: maxVal.toFixed(2), posZ };
}

function analyzeBeam() {
  const L = document.getElementById("length").value;
  const w1 = document.getElementById("w1").value;
  const w2 = document.getElementById("w2").value;
  const x = document.getElementById("distance").value;

  const beamData = new BeamData(L, w1, w2, x);

  const reactions = findMaxReactions(beamData);
  const maxBM = findMaxBM(beamData);
  const maxSF = findMaxSF(beamData);

  const bmList = [];
  for (let s = 0; s <= beamData.L; s++) {
    bmList.push(calculateBM(beamData, s));
  }

  const sfList = [];
  for (let t = 0; t <= beamData.L - beamData.x; t++) {
    sfList.push(calculateSF(beamData, t));
  }

  document.getElementById("reactionResults").innerText = `
        Maximum Rx_A: ${reactions.maxRxA} KN @ w1 at ${reactions.posA} m
        Maximum Rx_B: ${reactions.maxRxB} KN @ w1 at ${reactions.posB} m
    `;

  document.getElementById("sfResults").innerText = `
      
         BM List: ${bmList.join(", ")}
        SF List: ${sfList.join(", ")}
    `;
  document.getElementById("bmResults").innerText = `
    Maximum BM: ${maxBM.maxBM} kN-m @ y = ${maxBM.posY} m
      Maximum SF: ${maxSF.maxSF} kN @ z = ${maxSF.posZ} m
   
`;

  document.getElementById("results").style.display = "block";
}
